db/rdbms/src/main/java/io/camunda/db/rdbms/read/service/DecisionInstanceReader.java

```diff
@@ -0,0 +1,132 @@
+    return Optional.ofNullable(result.items()).flatMap(it -> it.stream().findFirst());
+  }
+
+  public SearchQueryResult<DecisionInstanceEntity> search(final DecisionInstanceQuery query) {
+    final var dbSort =
+        convertSort(query.sort(), DecisionInstanceSearchColumn.DECISION_INSTANCE_KEY);
+    final var dbQuery =
+        DecisionInstanceDbQuery.of(
+            b -> b.filter(query.filter()).sort(dbSort).page(convertPaging(dbSort, query.page())));
+
+    LOG.trace("[RDBMS DB] Search for process instance with filter {}", dbQuery);
+    final var totalHits = decisionInstanceMapper.count(dbQuery);
+    final var hits = enhanceEntities(decisionInstanceMapper.search(dbQuery), query.resultConfig());
+
+    return new SearchQueryResult<>(totalHits.intValue(), hits, extractSortValues(hits, dbSort));
+  }
+
+  /**
+   * Based on the result config, re batch-load here additional data (input, output values) with one
+   * SQL each (if enabled).
+   */
+  private List<DecisionInstanceEntity> enhanceEntities(
+      final List<DecisionInstanceEntity> intermediateResult,
+      final DecisionInstanceQueryResultConfig resultConfig) {
+    if (intermediateResult.isEmpty()) {
+      return intermediateResult;
+    }
+
+    if (resultConfig == null) {
+      return intermediateResult;
+    }
+
+    if (!resultConfig.includeEvaluatedInputs() && !resultConfig.includeEvaluatedOutputs()) {
+      return intermediateResult;
+    }
+
+    final Map<Long, List<EvaluatedInput>> inputs = new HashMap<>();
+    final Map<Long, List<EvaluatedOutput>> outputs = new HashMap<>();
+    if (resultConfig.includeEvaluatedInputs()) {
+      inputs.putAll(
+          decisionInstanceMapper
+              .loadInputs(intermediateResult.stream().map(DecisionInstanceEntity::key).toList())
+              .stream()
+              .collect(
+                  Collectors.groupingBy(
+                      DecisionInstanceDbModel.EvaluatedInput::decisionInstanceKey)));
+    }
+    if (resultConfig.includeEvaluatedOutputs()) {
+      outputs.putAll(
+          decisionInstanceMapper
+              .loadOutputs(intermediateResult.stream().map(DecisionInstanceEntity::key).toList())
+              .stream()
+              .collect(
+                  Collectors.groupingBy(
+                      DecisionInstanceDbModel.EvaluatedOutput::decisionInstanceKey)));
+    }
+
+    return intermediateResult.stream()
+        .map(
+            entity ->
+                entity.toBuilder()
+                    .evaluatedInputs(mapInputList(inputs.getOrDefault(entity.key(), List.of())))
+                    .evaluatedOutputs(mapOutputList(outputs.getOrDefault(entity.key(), List.of())))
+                    .build())
+        .toList();
+  }
+
+  private List<DecisionInstanceEntity.DecisionInstanceInputEntity> mapInputList(
+      final List<DecisionInstanceDbModel.EvaluatedInput> inputList) {
+    return inputList.stream()
```

Comment:
Didn't found a solution which was easier to read. Mapping before grouping is not possible because the resulting entity does not have the key (the initial reason for using the DbModel).